File: frontend/src/utils/callValidity.js
/**
 * Client-side invalid-call rules — mirrors backend call_validity.py.
 * API fields (sentiment_label, is_valid_call) take precedence when present.
 */

import { getCanonicalResult } from './canonicalResult';

const MIN_TRANSCRIPT_CHARS = 20;
const MIN_SUMMARY_CHARS = 15;
const LOW_CONFIDENCE_THRESHOLD = 0.15;

const INVALID_SENTIMENT_TOKENS = new Set([
  'invalid',
  'unclassified',
  'unknown',
  'n/a',
  'na',
  'none',
  'unclear',
  'insufficient',
]);

const NOISE_TRANSCRIPT_PATTERNS = [
  /^\[?\s*silence\s*\]?\.?$/i,
  /^no\s+speech/i,
  /^inaudible/i,
  /^unable\s+to\s+transcribe/i,
  /^transcription\s+failed/i,
  /^[\s.\-_,;:!?…]*$/,
];

function normalizeSentimentToken(sentiment) {
  if (!sentiment) return null;
  const value = String(sentiment).trim().toLowerCase();
  if (INVALID_SENTIMENT_TOKENS.has(value)) return null;
  if (value === 'positive' || value === 'pos') return 'positive';
  if (value === 'negative' || value === 'neg') return 'negative';
  if (value === 'neutral' || value === 'mixed' || value === 'neu') return 'neutral';
  return null;
}

function isNoiseTranscript(transcript) {
  const stripped = (transcript || '').trim();
  if (!stripped) return true;
  return NOISE_TRANSCRIPT_PATTERNS.some((pattern) => pattern.test(stripped));
}

function assessProviderResult(result) {
  if (!result || result.status !== 'completed') {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Analysis did not complete successfully',
    };
  }

  const transcript = (result.transcript || '').trim();
  if (!transcript) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Empty transcript — no speech content detected',
    };
  }

  if (transcript.length < MIN_TRANSCRIPT_CHARS) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Transcript too short to classify sentiment reliably',
    };
  }

  if (isNoiseTranscript(transcript)) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Transcript contains only silence or unusable audio content',
    };
  }

  const rawSentiment = (result.analysis?.sentiment || '').trim().toLowerCase();
  if (INVALID_SENTIMENT_TOKENS.has(rawSentiment)) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Insufficient content to infer sentiment',
    };
  }

  const sentimentLabel = normalizeSentimentToken(result.analysis?.sentiment);
  const summary = (result.analysis?.summary || '').trim();
  const confidence = result.analysis?.confidence ?? 0;

  if (!sentimentLabel) {
    if (confidence < LOW_CONFIDENCE_THRESHOLD && summary.length < MIN_SUMMARY_CHARS) {
      return {
        isValidCall: false,
        sentimentLabel: 'invalid',
        invalidReason: 'Low confidence and no useful summary — cannot classify reliably',
      };
    }
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Sentiment could not be classified from available content',
    };
  }

  if (confidence < LOW_CONFIDENCE_THRESHOLD && summary.length < MIN_SUMMARY_CHARS) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'Confidence too low with insufficient summary for reliable classification',
    };
  }

  return { isValidCall: true, sentimentLabel, invalidReason: null };
}

/** Derive validity from a recording when API fields are absent. */
export function deriveRecordingAssessment(record) {
  const aggregateStatus = record?.aggregate_status || record?.status || 'unknown';
  const resultsReady = record?.results_ready === true;
  const canonical = getCanonicalResult(record);

  if (!resultsReady) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: null,
    };
  }

  if (aggregateStatus === 'failed' && !canonical) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: record?.error || 'Analysis failed — no usable result produced',
    };
  }

  if (!canonical) {
    return {
      isValidCall: false,
      sentimentLabel: 'invalid',
      invalidReason: 'No successful analysis output from any solution',
    };
  }

  return assessProviderResult(canonical);
}

/**
 * Unified recording assessment — prefers API-normalized fields, falls back to client derivation.
 */
export function getRecordingAssessment(record) {
  if (record?.sentiment_label != null && record?.is_valid_call != null) {
    return {
      isValidCall: record.is_valid_call,
      sentimentLabel: record.sentiment_label,
      invalidReason: record.invalid_reason || null,
    };
  }
  return deriveRecordingAssessment(record);
}

export function getSentimentLabel(record) {
  return getRecordingAssessment(record).sentimentLabel;
}

export function isValidCall(record) {
  return getRecordingAssessment(record).isValidCall;
}

export function getInvalidReason(record) {
  return getRecordingAssessment(record).invalidReason;
}

File: backend/app/services/call_validity.py
"""Rules for classifying recordings as valid vs invalid for sentiment KPIs."""

from __future__ import annotations

import re
from dataclasses import dataclass

from app.models.schemas import ProviderResult

MIN_TRANSCRIPT_CHARS = 20
MIN_SUMMARY_CHARS = 15
LOW_CONFIDENCE_THRESHOLD = 0.15

INVALID_SENTIMENT_TOKENS = frozenset(
    {
        "invalid",
        "unclassified",
        "unknown",
        "n/a",
        "na",
        "none",
        "unclear",
        "insufficient",
    }
)

NOISE_TRANSCRIPT_PATTERNS = (
    re.compile(r"^\[?\s*silence\s*\]?\.?$", re.I),
    re.compile(r"^no\s+speech", re.I),
    re.compile(r"^inaudible", re.I),
    re.compile(r"^unable\s+to\s+transcribe", re.I),
    re.compile(r"^transcription\s+failed", re.I),
    re.compile(r"^[\s\.\-_,;:!?…]*$"),
)


@dataclass(frozen=True)
class RecordingValidity:
    is_valid_call: bool
    sentiment_label: str  # positive | neutral | negative | invalid
    invalid_reason: str | None = None


def _normalize_sentiment_token(sentiment: str | None) -> str | None:
    if not sentiment:
        return None
    value = sentiment.strip().lower()
    if value in INVALID_SENTIMENT_TOKENS:
        return None
    if value in {"positive", "pos"}:
        return "positive"
    if value in {"negative", "neg"}:
        return "negative"
    if value in {"neutral", "mixed", "neu"}:
        return "neutral"
    return None


def _is_noise_transcript(transcript: str) -> bool:
    stripped = transcript.strip()
    if not stripped:
        return True
    return any(pattern.match(stripped) for pattern in NOISE_TRANSCRIPT_PATTERNS)


def _assess_provider_result(result: ProviderResult) -> RecordingValidity:
    if result.status != "completed":
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Analysis did not complete successfully",
        )

    transcript = (result.transcript or "").strip()
    if not transcript:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Empty transcript — no speech content detected",
        )

    if len(transcript) < MIN_TRANSCRIPT_CHARS:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Transcript too short to classify sentiment reliably",
        )

    if _is_noise_transcript(transcript):
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Transcript contains only silence or unusable audio content",
        )

    raw_sentiment = (result.analysis.sentiment or "").strip().lower()
    if raw_sentiment in INVALID_SENTIMENT_TOKENS:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Insufficient content to infer sentiment",
        )

    sentiment_label = _normalize_sentiment_token(result.analysis.sentiment)
    summary = (result.analysis.summary or "").strip()
    confidence = result.analysis.confidence or 0.0

    if sentiment_label is None:
        if confidence < LOW_CONFIDENCE_THRESHOLD and len(summary) < MIN_SUMMARY_CHARS:
            return RecordingValidity(
                is_valid_call=False,
                sentiment_label="invalid",
                invalid_reason="Low confidence and no useful summary — cannot classify reliably",
            )
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Sentiment could not be classified from available content",
        )

    if confidence < LOW_CONFIDENCE_THRESHOLD and len(summary) < MIN_SUMMARY_CHARS:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="Confidence too low with insufficient summary for reliable classification",
        )

    return RecordingValidity(is_valid_call=True, sentiment_label=sentiment_label)


def assess_recording_validity(
    *,
    aggregate_status: str,
    job_error: str | None,
    canonical: ProviderResult | None,
    results_ready: bool,
) -> RecordingValidity:
    """Assess one recording using its canonical (winner) solution output."""
    if not results_ready:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason=None,
        )

    if aggregate_status == "failed" and canonical is None:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason=job_error or "Analysis failed — no usable result produced",
        )

    if canonical is None:
        return RecordingValidity(
            is_valid_call=False,
            sentiment_label="invalid",
            invalid_reason="No successful analysis output from any solution",
        )

    return _assess_provider_result(canonical)
